### ISSUE TITLE
Downloading results as xlsx returns an invalid file when disk is full

### ISSUE THREAD

**Issue by markberger** [metabase/metabase#13159]
**Describe the bug**
When disk is full and the user requests to download a xlsx file, Metabase returns an invalid file.

**To Reproduce**
Run Metabase on an instance where the disk is full or mock a disk is full error. Try to download the results of a question in xlsx. Excel will not load the file. If you open the file with Google Sheets you will something like the following:
```
{"database_id":35 | started_at:"2020-08-24T18:10:56.594Z" | via:[{"status":"failed" | class:"class java.lang.RuntimeException" | error:"java.io.IOException: No space left on device" | stacktrace: ... }
```

**Expected behavior**
Metabase should tell the user it was unable to generate a xlsx file. Ideally it would mention that the disk is full.

**Information about your Metabase Installation:**
- Your browser and the version: Chrome 84.0.4147.135 
- Your operating system: macOS 10.15.6 
- Your databases: BigQuery
- Metabase version: v1.35.4.1
- Metabase hosting environment: Ubuntu
- Metabase internal database: Postgres

**Severity**
Annoying / bad user experience

**Comment by flamber** [comment]
Would probably require a dialog like #10566, since I don't think it's possible to change the streaming HTTP code.

**Comment by camsaul** [comment]
I think this should be fixed in 0.36.0+ where we switched to a streaming implementation in #12521. I'm guessing the disk usage had to happen for really large result sets, but shouldn't be an issue anymore since we basically write the file directly to the HTTP response as results come in

**Comment by camsaul** [comment]
This is probably already fixed, but we should confirm

**Comment by noahmoss** [comment]
I don't think this is fixed, actually. From what I can tell, the problem is that SXSSFWorkbook doesn't actually write the export to the stream incrementally -- it actually writes it to a temp file on disk, and then we write the temp file to the stream when the export has been fully generated. The "streaming" aspect of SXSSFWorkbook refers to the fact that it doesn't store the workbook in memory, thus avoiding OOM issues, but it still needs an adequate amount of disk space. I don't see any way around that.

**Comment by noahmoss** [comment]
Here's how I tested it locally, on Mac OS:
1. Run `ulimit -f 1000` to restrict the size limit of files written by processes spawned by the current shell. Found this value through trial and error; might be different on other machines with different file block sizes.
2. Start Metabase in the same shell
3. Download an XLSX file of a small dataset (i.e. filter a table in the sample database to just one row). It should succeed.
4. Download an XLSX file of a large dataset (i.e. the entire `Orders` table). The resulting file should be invalid and there will be a `File too large` exception in the logs.

I think this is basically an inherent limitation for generating XLSX files. You need enough disk space to write the entire temp file while it's being generated. So in that sense, we can't solve this issue as currently stated.

However since we're not actually streaming the XLSX results to the caller as they're being generated, we _could_ switch to using a non-streaming endpoint, and then show an error message on the frontend if an error occurs during export generation. (But we can't do this for CSV/JSON since we actually _are_ streaming results for them, so there would be a weird inconsistency there).

In addition, #19480 was likely underlying issue that was leading to disk space issues in the first place. So this may not be a problem in practice any more.

So in conclusion, this isn't technically fixed, but I'm not sure if this issue should remain open.

**Comment by flamber** [comment]
@noahmoss Good writeup - I doubt anyone will hit this after #19480 was fixed.

We can close this issue, but then we need to open another about streaming the stacktrace.

Generally it's a much bigger problem with XLSX ([example](https://discourse.metabase.com/t/excel-files-not-exporting-properly/19516)), since it's a lot more often stacktraces happens here, but Excel also complicates it by not even showing the content (just shows an error "This file is invalid", so only people that open the file in Notepad can see the problem)
